mypy/semanal.py

```diff
@@ -432,51 +432,57 @@ def visit_overloaded_func_def(self, defn: OverloadedFuncDef) -> None:
         # Decide whether to analyze this as a property or an overload.  If an
         # overload, and we're outside a stub, find the impl and set it.  Remove
         # the impl from the item list, it's special.
-        t = []  # type: List[CallableType]
+        types = []  # type: List[CallableType]
         non_overload_indexes = []
-        for i, item in enumerate(defn.items):
-            # TODO support decorated overloaded functions properly
-            item.is_overload = True
-            item.accept(self)
 
-            if isinstance(item, Decorator):
-                item.func.is_overload = True
-                callable = function_type(item.func, self.builtin_type('builtins.function'))
-                assert isinstance(callable, CallableType)
-                if item.func.is_property and i == 0:
-                    # This defines a property, probably with a setter and/or
-                    # deleter.
-                    t.append(callable)
-                    self.analyze_property_with_multi_part_definition(defn)
-                    break
-                elif not any(refers_to_fullname(dec, 'typing.overload')
-                             for dec in item.decorators):
+        # See if the first item is a property (and not an overload)
+        first_item = defn.items[0]
+        first_item.is_overload = True
+        first_item.accept(self)
+
+        if isinstance(first_item, Decorator) and first_item.func.is_property:
+            first_item.func.is_overload = True
+            self.analyze_property_with_multi_part_definition(defn)
+            typ = function_type(first_item.func, self.builtin_type('builtins.function'))
+            assert isinstance(typ, CallableType)
+            types = [typ]
+        else:
+            for i, item in enumerate(defn.items):
+                if i != 0:
+                    # The first item was already visited
+                    item.is_overload = True
+                    item.accept(self)
+                # TODO support decorated overloaded functions properly
+                if isinstance(item, Decorator):
+                    item.func.is_overload = True
```

Comment:
But then why is it not set for non-decorated implementations? Something's still fishy.